relay-server/src/actors/relays.rs

```diff
@@ -208,165 +285,39 @@ impl RelayCache {
-    }
-}
-
-impl Supervised for RelayCache {}
-
-impl SystemService for RelayCache {}
-
-impl Default for RelayCache {
-    fn default() -> Self {
-        unimplemented!("register with the SystemRegistry instead")
-    }
-}
-
-#[derive(Debug)]
-pub struct GetRelay {
-    pub relay_id: RelayId,
-}
-
-#[derive(Debug)]
-pub struct GetRelayResult {
-    pub relay: Option<RelayInfo>,
-}
-
-impl Message for GetRelay {
-    type Result = Result<GetRelayResult, KeyError>;
-}
-
-impl Handler<GetRelay> for RelayCache {
-    type Result = Response<GetRelayResult, KeyError>;
-
-    fn handle(&mut self, message: GetRelay, context: &mut Self::Context) -> Self::Result {
-        self.get_or_fetch_info(message.relay_id, context)
-            .map(|(_id, relay)| GetRelayResult { relay })
+            .or_insert_with(Vec::new)
+            .push(sender);
     }
 }
 
-#[derive(Debug, Deserialize, Serialize)]
-pub struct GetRelays {
-    pub relay_ids: Vec<RelayId>,
-}
+impl Service for RelayCacheService {
+    type Interface = RelayCache;
 
-#[derive(Debug, Serialize, Deserialize)]
-pub struct GetRelaysResult {
-    /// new format public key plus additional parameters
-    pub relays: HashMap<RelayId, Option<RelayInfo>>,
-}
+    fn spawn_handler(mut self, mut rx: relay_system::Receiver<Self::Interface>) {
+        tokio::spawn(async move {
+            relay_log::info!("key cache started");
 
-impl From<PublicKeysResultCompatibility> for GetRelaysResult {
-    fn from(relays_info: PublicKeysResultCompatibility) -> Self {
-        let relays = if relays_info.relays.is_empty() && !relays_info.public_keys.is_empty() {
-            relays_info
-                .public_keys
-                .into_iter()
-                .map(|(id, pk)| (id, pk.map(RelayInfo::new)))
-                .collect()
-        } else {
-            relays_info.relays
-        };
-        Self { relays }
-    }
-}
-
-/// Defines a compatibility format for deserializing relays info that supports
-/// both the old and the new format for relay info
-#[derive(Debug, Deserialize)]
-#[serde(rename_all = "camelCase")]
-pub struct PublicKeysResultCompatibility {
-    /// DEPRECATED. Legacy format only public key info.
-    #[serde(default, rename = "public_keys")]
-    pub public_keys: HashMap<RelayId, Option<PublicKey>>,
-    /// A map from Relay's identifier to its information.
-    #[serde(default)]
-    pub relays: HashMap<RelayId, Option<RelayInfo>>,
-}
-
-impl Message for GetRelays {
-    type Result = Result<GetRelaysResult, KeyError>;
-}
-
-impl UpstreamQuery for GetRelays {
-    type Response = PublicKeysResultCompatibility;
-
-    fn method(&self) -> Method {
-        Method::POST
-    }
-
-    fn path(&self) -> Cow<'static, str> {
-        Cow::Borrowed("/api/0/relays/publickeys/")
-    }
-
-    fn priority() -> RequestPriority {
-        RequestPriority::High
-    }
-
-    fn retry() -> bool {
-        false
-    }
-}
-
-impl Handler<GetRelays> for RelayCache {
```

Comment:
This has been moved to the call site in the endpoint handler, allowing us to avoid spawning yet another task over here.